### DISCUSSION TITLE
Is it possible to have two different sections of frozen bytecode?

### DISCUSSION THREAD

**Discussion by victorallume** [micropython/micropython#11647] [Core Development] [answered]
Is it possible/feasible to have two different sections of flash containing frozen bytecode, but one is loaded at a later point (i.e. at some point after boot.py has been run)?

**Comment by jimmo** [comment]
@victorallume A mostly-accurate way to think of frozen bytecode is as a special filesystem containing .mpy files where the loader knows that certain data structures are (a) already laid out exactly as the runtime expects, and (b) directly memory mappable.

Compare to loading a .mpy file from the filesystem, we need to do more translation of the data structures (because the .mpy format is generic whereas the in-memory structure the runtime expects depends on the architecture), although as of the v1.19 there is much less translation. But from the filesystem you cannot do memory mapping because the files are not necessarily contiguous.

Anyway, the point is that none of this happens until you import a given module. So unless I've misunderstood the question, I don't think you need to worry about this?

**Reply by victorallume** [reply]
I'm looking at having two separate sections/stages of frozen bytecode (not on a filesystem). One of them is loaded first, and contains some low level operations, e.g. networking and OTA. The first stage would verify the second stage, then start running it. I'm currently doing this with the second stage living on a filesystem in .mpy files, but in the interests of saving RAM, would like to get rid of the filesystem part. Is that a bit clearer?

**Reply by jimmo** [reply]
Yep, I'm just not clear what the problem is, as each frozen module is independent of any other module in terms of how it's loaded.

i.e. If I have `foo.py`, `bar.py`, `baz.py` all frozen, I can `import foo` (and assuming that `foo.py` doesn't import bar or baz) then no loading of bar or baz will happen.

I guess what I'm getting at is that I'm interpreting your question as "I want to isolate loading of the first stage modules from the second stage modules, i.e. I don't want to incur any cost of the second stage modules until I'm ready", but that is exactly how it works anyway.

I am a bit unsure what you mean by "The first stage would verify the second stage" though.

You may also be interested in what Damien is experimenting with in https://github.com/micropython/micropython/pull/8381 (but I think this is likely to be a "MicroPython 2.0" feature).

**Reply by victorallume** [reply]
The problem is not about cost of loading the second stage modules, and more about verifying the second stage, and allowing the second stage to be updated independently of the first. 

Regarding verification: this checksums the flash region for the relevant stage, compares against a known checksum, and if it fails, goes into a 'safe mode' and attempts an OTA.

**Reply by peterhinch** [reply]
I'm pretty sure the flashing process verifies the outcome so runtime checks are not necessary.

**Reply by victorallume** [reply]
Thanks @peterhinch - we're flashing via an OTA (i.e. from micropython code), so would like an extra level of verification; we also perform this on boot as an extra check. But if we put that aside, the other part is to be able to update the two sections of bytecode independently.

**Reply by jimmo** [reply, marked as answer]
@victorallume I think the thing you want is https://github.com/micropython/micropython/pull/8381 :D

(This feature is going to be a pretty big deal for many people I think, because it gives you almost all the benefits of frozen code but with all the advantages of not having to build and flash the firmware. It's also really great for development -- see https://github.com/micropython/micropython/issues/8426#issuecomment-1276997914 -- because you can do a full deploy of your entire codebase much much faster than copying .py/.mpy files to the regular filesystem).

But in terms of what you can do now.... I think I understand your original question now. It's not so much that you want two different groups of frozen code to be loaded independently (they are anyway), rather that you plan to load them independently, so you'd also like them to be physically laid out independently on the flash (e.g. so you could update just one region via OTA or something).

In theory you can use the linker script to put particular symbols into different regions, however there is a bunch of stuff in the frozen code section that is shared (for example the QSTR table), and we definitely don't have a way to safely only update a subsection of it.

**Reply by victorallume** [reply]
Thanks @jimmo - that PR does indeed look like what we're after.
